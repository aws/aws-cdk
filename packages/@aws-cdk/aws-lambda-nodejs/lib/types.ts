import { BundlingDockerImage } from '@aws-cdk/core';

/**
 * Bundling options
 */
export interface BundlingOptions {
  /**
   * Whether to minify files when bundling.
   *
   * @default false
   */
  readonly minify?: boolean;

  /**
   * Whether to include source maps when bundling.
   *
   * @default false
   */
  readonly sourceMap?: boolean;

  /**
   * Target environment for the generated JavaScript code.
   *
   * @see https://esbuild.github.io/api/#target
   *
   * @default - the node version of the runtime
   */
  readonly target?: string;

  /**
   * Use loaders to change how a given input file is interpreted.
   *
   * Configuring a loader for a given file type lets you load that file type with
   * an `import` statement or a `require` call.
   *
   * @see https://esbuild.github.io/api/#loader
   *
   * @example { '.png': 'dataurl' }
   *
   * @default - use esbuild default loaders
   */
  readonly loader?: { [ext: string]: string };

  /**
   * Log level for esbuild
   *
   * @default LogLevel.WARNING
   */
  readonly logLevel?: LogLevel;

  /**
   * Whether to preserve the original `name` values even in minified code.
   *
   * In JavaScript the `name` property on functions and classes defaults to a
   * nearby identifier in the source code.
   *
   * However, minification renames symbols to reduce code size and bundling
   * sometimes need to rename symbols to avoid collisions. That changes value of
   * the `name` property for many of these cases. This is usually fine because
   * the `name` property is normally only used for debugging. However, some
   * frameworks rely on the `name` property for registration and binding purposes.
   * If this is the case, you can enable this option to preserve the original
   * `name` values even in minified code.
   *
   * @default false
   */
  readonly keepNames?: boolean;

  /**
   * Normally the esbuild automatically discovers `tsconfig.json` files and reads their contents during a build.
   *
   * However, you can also configure a custom `tsconfig.json` file to use instead.
   *
   * This is similar to entry path, you need to provide path to your custom `tsconfig.json`.
   *
   * This can be useful if you need to do multiple builds of the same code with different settings.
   *
   * @example { 'tsconfig': 'path/custom.tsconfig.json' }
   *
   * @default - automatically discovered by `esbuild`
   */
  readonly tsconfig? : string

  /**
   * This option tells esbuild to write out a JSON file relative to output directory with metadata about the build.
   *
   * The metadata in this JSON file follows this schema (specified using TypeScript syntax):
   *
   * ```typescript
   *  {
   *     outputs: {
   *          [path: string]: {
   *            bytes: number
   *            inputs: {
   *              [path: string]: { bytesInOutput: number }
   *            }
   *            imports: { path: string }[]
   *            exports: string[]
   *          }
   *        }
   *     }
   * }
   * ```
   * This data can then be analyzed by other tools. For example,
   * bundle buddy can consume esbuild's metadata format and generates a treemap visualization
   * of the modules in your bundle and how much space each one takes up.
   * @see https://esbuild.github.io/api/#metafile
   * @default - false
   */
  readonly metafile?: boolean

  /**
   * Use this to insert an arbitrary string at the beginning of generated JavaScript files.
   *
   * This is similar to footer which inserts at the end instead of the beginning.
   *
   * This is commonly used to insert comments:
   *
   * @default -  no comments are passed
   */
  readonly banner? : string

  /**
   * Use this to insert an arbitrary string at the end of generated JavaScript files.
   *
   * This is similar to banner which inserts at the beginning instead of the end.
   *
   * This is commonly used to insert comments
   *
   * @default -  no comments are passed
   */
  readonly footer? : string

  /**
   * Environment variables defined when bundling runs.
   *
   * @default - no environment variables are defined.
   */
  readonly environment?: { [key: string]: string; };

  /**
   * Replace global identifiers with constant expressions.
   *
   * @example { 'process.env.DEBUG': 'true' }
   * @example { 'process.env.API_KEY': JSON.stringify('xxx-xxxx-xxx') }
   *
   * @default - no replacements are made
   */
  readonly define?: { [key: string]: string };

  /**
   * A list of modules that should be considered as externals (already available
   * in the runtime).
   *
   * @default ['aws-sdk']
   */
  readonly externalModules?: string[];

  /**
   * A list of modules that should be installed instead of bundled. Modules are
   * installed in a Lambda compatible environment only when bundling runs in
   * Docker.
   *
   * @default - all modules are bundled
   */
  readonly nodeModules?: string[];

  /**
   * The version of esbuild to use when running in a Docker container.
   *
   * @default - latest v0
   */
  readonly esbuildVersion?: string;

  /**
   * Build arguments to pass when building the bundling image.
   *
   * @default - no build arguments are passed
   */
  readonly buildArgs?: { [key:string] : string };

  /**
   * Force bundling in a Docker container even if local bundling is
   * possible. This is useful if your function relies on node modules
   * that should be installed (`nodeModules`) in a Lambda compatible
   * environment.
   *
   * @default false
   */
  readonly forceDockerBundling?: boolean;

  /**
   * A custom bundling Docker image.
   *
   * This image should have esbuild installed globally. If you plan to use `nodeModules`
   * it should also have `npm` or `yarn` depending on the lock file you're using.
   *
   * See https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda-nodejs/lib/Dockerfile
   * for the default image provided by @aws-cdk/aws-lambda-nodejs.
   *
   * @default - use the Docker image provided by @aws-cdk/aws-lambda-nodejs
   */
  readonly dockerImage?: BundlingDockerImage;

  /**
   * Command hooks
   *
   * @default - do not run additional commands
   */
  readonly commandHooks?: ICommandHooks;
}

/**
 * Command hooks
 *
 * These commands will run in the environment in which bundling occurs: inside
 * the container for Docker bundling or on the host OS for local bundling.
 *
 * Commands are chained with `&&`.
 *
 * @example
 * {
 *   // Copy a file from the input directory to the output directory
 *   // to include it in the bundled asset
 *   afterBundling(inputDir: string, outputDir: string): string[] {
 *     return [`cp ${inputDir}/my-binary.node ${outputDir}`];
 *   }
 *   // ...
 * }
 */
export interface ICommandHooks {
  /**
   * Returns commands to run before bundling.
   *
   * Commands are chained with `&&`.
   */
  beforeBundling(inputDir: string, outputDir: string): string[];

  /**
   * Returns commands to run before installing node modules.
   *
   * This hook only runs when node modules are installed.
   *
   * Commands are chained with `&&`.
   */
  beforeInstall(inputDir: string, outputDir: string): string[];

  /**
   * Returns commands to run after bundling.
   *
   * Commands are chained with `&&`.
   */
  afterBundling(inputDir: string, outputDir: string): string[];
}

/**
 * Log level for esbuild
 */
export enum LogLevel {
  /** Show everything */
  INFO = 'info',
  /** Show warnings and errors */
  WARNING = 'warning',
  /** Show errors only */
  ERROR = 'error',
  /** Show nothing */
  SILENT = 'silent',
}
