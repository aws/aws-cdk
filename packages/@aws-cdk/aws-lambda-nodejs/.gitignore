*.js
tsconfig.json
*.js.map
*.d.ts
*.generated.ts
dist
lib/generated/resources.ts
.jsii

.LAST_BUILD
.nyc_output
coverage
nyc.config.js
.LAST_PACKAGE
*.snk

# Parcel
.build
.cache

!test/integ-handlers/js-handler.js
!test/function.test.handler2.js
!.eslintrc.js
