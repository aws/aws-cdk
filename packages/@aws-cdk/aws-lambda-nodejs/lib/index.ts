export * from './function';
export * from './bundling';
