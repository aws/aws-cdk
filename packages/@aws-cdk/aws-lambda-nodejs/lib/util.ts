import * as fs from 'fs';
import * as path from 'path';

// From https://github.com/errwischt/stacktrace-parser/blob/master/src/stack-trace-parser.js
const STACK_RE = /^\s*at (?:((?:\[object object\])?[^\\/]+(?: \[as \S+\])?) )?\(?(.*?):(\d+)(?::(\d+))?\)?\s*$/i;

/**
 * A parsed stack trace line
 */
export interface StackTrace {
  readonly file: string;
  readonly methodName?: string;
  readonly lineNumber: number;
  readonly column: number;
}

/**
 * Parses the stack trace of an error
 */
export function parseStackTrace(error?: Error): StackTrace[] {
  const err = error || new Error();

  if (!err.stack) {
    return [];
  }

  const lines = err.stack.split('\n');

  const stackTrace: StackTrace[] = [];

  for (const line of lines) {
    const results = STACK_RE.exec(line);
    if (results) {
      stackTrace.push({
        file: results[2],
        methodName: results[1],
        lineNumber: parseInt(results[3], 10),
        column: parseInt(results[4], 10),
      });
    }
  }

  return stackTrace;
}

/**
 * Returns the major version of node installation
 */
export function nodeMajorVersion(): number {
  return parseInt(process.versions.node.split('.')[0], 10);
}

/**
 * Finds the closest path containg a path
 */
function findClosestPathContaining(p: string): string | undefined {
  for (const nodeModulesPath of module.paths) {
    if (fs.existsSync(path.join(path.dirname(nodeModulesPath), p))) {
      return path.dirname(nodeModulesPath);
    }
  }

  return undefined;
}

/**
 * Finds closest package.json path
 */
export function findPkgPath(): string | undefined {
  return findClosestPathContaining('package.json');
}

/**
 * Finds closest .git/
 */
export function findGitPath(): string | undefined {
  return findClosestPathContaining(`.git${path.sep}`);
}
