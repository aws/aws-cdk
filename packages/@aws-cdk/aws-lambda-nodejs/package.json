{
  "name": "@aws-cdk/aws-lambda-nodejs",
  "version": "0.0.0",
  "description": "The CDK Construct Library for AWS Lambda in Node.js",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "jsii": {
    "outdir": "dist",
    "targets": {
      "java": {
        "package": "software.amazon.awscdk.services.lambda.nodejs",
        "maven": {
          "groupId": "software.amazon.awscdk",
          "artifactId": "lambda-nodejs"
        }
      },
      "dotnet": {
        "namespace": "Amazon.CDK.AWS.Lambda.Nodejs",
        "packageId": "Amazon.CDK.AWS.Lambda.Nodejs",
        "iconUrl": "https://raw.githubusercontent.com/aws/aws-cdk/master/logo/default-256-dark.png"
      },
      "python": {
        "distName": "aws-cdk.aws-lambda-nodejs",
        "module": "aws_cdk.aws_lambda_nodejs",
        "classifiers": [
          "Framework :: AWS CDK",
          "Framework :: AWS CDK :: 1"
        ]
      }
    },
    "projectReferences": true
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/aws/aws-cdk.git",
    "directory": "packages/@aws-cdk/aws-lambda-nodejs"
  },
  "scripts": {
    "build": "cdk-build",
    "watch": "cdk-watch",
    "lint": "cdk-lint",
    "test": "cdk-test",
    "integ": "cdk-integ",
    "pkglint": "pkglint -f",
    "package": "cdk-package",
    "awslint": "cdk-awslint",
    "build+test+package": "yarn build+test && yarn package",
    "build+test": "yarn build && yarn test",
    "compat": "cdk-compat",
    "rosetta:extract": "yarn --silent jsii-rosetta extract",
    "build+extract": "yarn build && yarn rosetta:extract",
    "build+test+extract": "yarn build+test && yarn rosetta:extract"
  },
  "keywords": [
    "aws",
    "cdk",
    "constructs",
    "lambda"
  ],
  "author": {
    "name": "Amazon Web Services",
    "url": "https://aws.amazon.com",
    "organization": true
  },
  "license": "Apache-2.0",
  "devDependencies": {
    "@types/jest": "^26.0.23",
    "@aws-cdk/aws-ec2": "0.0.0",
    "cdk-build-tools": "0.0.0",
    "cdk-integ-tools": "0.0.0",
    "delay": "5.0.0",
    "esbuild": "^0.12.9",
    "pkglint": "0.0.0",
    "@aws-cdk/assert-internal": "0.0.0"
  },
  "dependencies": {
    "@aws-cdk/aws-lambda": "0.0.0",
    "@aws-cdk/core": "0.0.0",
    "constructs": "^3.3.69"
  },
  "homepage": "https://github.com/aws/aws-cdk",
  "peerDependencies": {
    "@aws-cdk/aws-lambda": "0.0.0",
    "@aws-cdk/core": "0.0.0",
    "constructs": "^3.3.69"
  },
  "engines": {
    "node": ">= 10.13.0 <13 || >=13.7.0"
  },
  "stability": "stable",
  "maturity": "stable",
  "awscdkio": {
    "announce": false
  },
  "nozem": false,
  "cdk-build": {
    "jest": true
  },
  "publishConfig": {
    "tag": "latest"
  }
}
