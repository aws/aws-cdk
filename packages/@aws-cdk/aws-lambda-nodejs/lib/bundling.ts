import * as os from 'os';
import * as path from 'path';
import { AssetCode, Code, Runtime } from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';
import { BundlingOptions } from './types';
import { exec, extractDependencies, findUp, getEsBuildVersion, LockFile } from './util';

const ESBUILD_VERSION = '0';

/**
 * Bundling properties
 */
export interface BundlingProps extends BundlingOptions {
  /**
   * Path to lock file
   */
  readonly depsLockFilePath: string;

  /**
   * Entry file
   */
  readonly entry: string;

  /**
   * The runtime of the lambda function
   */
  readonly runtime: Runtime;
}

/**
 * Bundling with esbuild
 */
export class Bundling implements cdk.BundlingOptions {
  /**
   * esbuild bundled Lambda asset code
   */
  public static bundle(options: BundlingProps): AssetCode {
    return Code.fromAsset(path.dirname(options.depsLockFilePath), {
      assetHashType: cdk.AssetHashType.OUTPUT,
      bundling: new Bundling(options),
    });
  }

  public static clearRunsLocallyCache(): void {
    this.runsLocally = undefined;
  }

  private static runsLocally?: boolean;

  // Core bundling options
  public readonly image: cdk.DockerImage;
  public readonly command: string[];
  public readonly environment?: { [key: string]: string };
  public readonly workingDirectory: string;
  public readonly local?: cdk.ILocalBundling;

  private readonly relativeEntryPath: string;
  private readonly relativeTsconfigPath?: string;
  private readonly externals: string[];

  constructor(private readonly props: BundlingProps) {
    Bundling.runsLocally = Bundling.runsLocally
      ?? getEsBuildVersion()?.startsWith(ESBUILD_VERSION)
      ?? false;

    const projectRoot = path.dirname(props.depsLockFilePath);
    this.relativeEntryPath = path.relative(projectRoot, path.resolve(props.entry));

    if (props.tsconfig) {
      this.relativeTsconfigPath = path.relative(projectRoot, path.resolve(props.tsconfig));
    }

    this.externals = [
      ...props.externalModules ?? ['aws-sdk'], // Mark aws-sdk as external by default (available in the runtime)
      ...props.nodeModules ?? [], // Mark the modules that we are going to install as externals also
    ];

    // Docker bundling
    const shouldBuildImage = props.forceDockerBundling || !Bundling.runsLocally;
    this.image = shouldBuildImage
      ? props.dockerImage ?? cdk.DockerImage.fromBuild(path.join(__dirname, '../lib'), {
        buildArgs: {
          ...props.buildArgs ?? {},
          IMAGE: props.runtime.bundlingDockerImage.image,
          ESBUILD_VERSION: props.esbuildVersion ?? ESBUILD_VERSION,
        },
      })
      : cdk.DockerImage.fromRegistry('dummy'); // Do not build if we don't need to

    const bundlingCommand = this.createBundlingCommand(cdk.AssetStaging.BUNDLING_INPUT_DIR, cdk.AssetStaging.BUNDLING_OUTPUT_DIR);
    this.command = ['bash', '-c', bundlingCommand];
    this.environment = props.environment;
    // Bundling sets the working directory to cdk.AssetStaging.BUNDLING_INPUT_DIR
    // and we want to force npx to use the globally installed esbuild.
    this.workingDirectory = '/';

    // Local bundling
    if (!props.forceDockerBundling) { // only if Docker is not forced
      const osPlatform = os.platform();
      const createLocalCommand = (outputDir: string) => this.createBundlingCommand(projectRoot, outputDir, osPlatform);

      this.local = {
        tryBundle(outputDir: string) {
          if (Bundling.runsLocally === false) {
            process.stderr.write('esbuild cannot run locally. Switching to Docker bundling.\n');
            return false;
          }

          const localCommand = createLocalCommand(outputDir);

          exec(
            osPlatform === 'win32' ? 'cmd' : 'bash',
            [
              osPlatform === 'win32' ? '/c' : '-c',
              localCommand,
            ],
            {
              env: { ...process.env, ...props.environment ?? {} },
              stdio: [ // show output
                'ignore', // ignore stdio
                process.stderr, // redirect stdout to stderr
                'inherit', // inherit stderr
              ],
              cwd: path.dirname(props.entry),
              windowsVerbatimArguments: osPlatform === 'win32',
            });

          return true;
        },
      };
    }
  }

  public createBundlingCommand(inputDir: string, outputDir: string, osPlatform: NodeJS.Platform = 'linux'): string {
    const pathJoin = osPathJoin(osPlatform);

    const npx = osPlatform === 'win32' ? 'npx.cmd' : 'npx';
    const loaders = Object.entries(this.props.loader ?? {});
    const defines = Object.entries(this.props.define ?? {});

    const esbuildCommand: string = [
      npx, 'esbuild',
      '--bundle', `"${pathJoin(inputDir, this.relativeEntryPath)}"`,
      `--target=${this.props.target ?? toTarget(this.props.runtime)}`,
      '--platform=node',
      `--outfile="${pathJoin(outputDir, 'index.js')}"`,
      ...this.props.minify ? ['--minify'] : [],
      ...this.props.sourceMap ? ['--sourcemap'] : [],
      ...this.externals.map(external => `--external:${external}`),
      ...loaders.map(([ext, name]) => `--loader:${ext}=${name}`),
      ...defines.map(([key, value]) => `--define:${key}=${value}`),
      ...this.props.logLevel ? [`--log-level=${this.props.logLevel}`] : [],
      ...this.props.keepNames ? ['--keep-names'] : [],
      ...this.relativeTsconfigPath ? [`--tsconfig=${pathJoin(inputDir, this.relativeTsconfigPath)}`] : [],
      ...this.props.metafile ? [`--metafile=${pathJoin(outputDir, 'index.meta.json')}`] : [],
      ...this.props.banner ? [`--banner='${this.props.banner}'`] : [],
      ...this.props.footer ? [`--footer='${this.props.footer}'`] : [],
    ].join(' ');

    let depsCommand = '';
    if (this.props.nodeModules) {
      // Find 'package.json' closest to entry folder, we are going to extract the
      // modules versions from it.
      const pkgPath = findUp('package.json', path.dirname(this.props.entry));
      if (!pkgPath) {
        throw new Error('Cannot find a `package.json` in this project. Using `nodeModules` requires a `package.json`.');
      }

      // Determine dependencies versions, lock file and installer
      const dependencies = extractDependencies(pkgPath, this.props.nodeModules);
      let installer = Installer.NPM;
      let lockFile = LockFile.NPM;
      if (this.props.depsLockFilePath.endsWith(LockFile.YARN)) {
        lockFile = LockFile.YARN;
        installer = Installer.YARN;
      }

      const osCommand = new OsCommand(osPlatform);

      // Create dummy package.json, copy lock file if any and then install
      depsCommand = chain([
        osCommand.writeJson(pathJoin(outputDir, 'package.json'), { dependencies }),
        osCommand.copy(pathJoin(inputDir, lockFile), pathJoin(outputDir, lockFile)),
        osCommand.changeDirectory(outputDir),
        `${installer} install`,
      ]);
    }

    return chain([
      ...this.props.commandHooks?.beforeBundling(inputDir, outputDir) ?? [],
      esbuildCommand,
      ...(this.props.nodeModules && this.props.commandHooks?.beforeInstall(inputDir, outputDir)) ?? [],
      depsCommand,
      ...this.props.commandHooks?.afterBundling(inputDir, outputDir) ?? [],
    ]);
  }
}

enum Installer {
  NPM = 'npm',
  YARN = 'yarn',
}

/**
 * OS agnostic command
 */
class OsCommand {
  constructor(private readonly osPlatform: NodeJS.Platform) {}

  public writeJson(filePath: string, data: any): string {
    const stringifiedData = JSON.stringify(data);
    if (this.osPlatform === 'win32') {
      return `echo ^${stringifiedData}^ > ${filePath}`;
    }

    return `echo '${stringifiedData}' > ${filePath}`;
  }

  public copy(src: string, dest: string): string {
    if (this.osPlatform === 'win32') {
      return `copy ${src} ${dest}`;
    }

    return `cp ${src} ${dest}`;
  }

  public changeDirectory(dir: string): string {
    return `cd ${dir}`;
  }
}

/**
 * Chain commands
 */
function chain(commands: string[]): string {
  return commands.filter(c => !!c).join(' && ');
}

/**
 * Platform specific path join
 */
function osPathJoin(platform: NodeJS.Platform) {
  return function(...paths: string[]): string {
    const joined = path.join(...paths);
    // If we are on win32 but need posix style paths
    if (os.platform() === 'win32' && platform !== 'win32') {
      return joined.replace(/\\/g, '/');
    }
    return joined;
  };
}

/**
 * Converts a runtime to an esbuild node target
 */
function toTarget(runtime: Runtime): string {
  const match = runtime.name.match(/nodejs(\d+)/);

  if (!match) {
    throw new Error('Cannot extract version from runtime.');
  }

  return `node${match[1]}`;
}
