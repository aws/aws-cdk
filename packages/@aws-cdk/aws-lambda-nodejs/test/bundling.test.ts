import * as child_process from 'child_process';
import * as os from 'os';
import * as path from 'path';
import { Code, Runtime } from '@aws-cdk/aws-lambda';
import { AssetHashType, DockerImage } from '@aws-cdk/core';
import { version as delayVersion } from 'delay/package.json';
import { Bundling } from '../lib/bundling';
import { LogLevel } from '../lib/types';
import * as util from '../lib/util';

jest.mock('@aws-cdk/aws-lambda');

// Mock BundlingDockerImage.fromAsset() to avoid building the image
let fromAssetMock = jest.spyOn(DockerImage, 'fromBuild');
let getEsBuildVersionMock = jest.spyOn(util, 'getEsBuildVersion');
beforeEach(() => {
  jest.clearAllMocks();
  jest.resetAllMocks();
  Bundling.clearRunsLocallyCache();
  getEsBuildVersionMock.mockReturnValue('0.8.8');
  fromAssetMock.mockReturnValue({
    image: 'built-image',
    cp: () => 'dest-path',
    run: () => {},
    toJSON: () => 'built-image',
  });
});

let depsLockFilePath = '/project/yarn.lock';
let entry = '/project/lib/handler.ts';
let tsconfig = '/project/lib/custom-tsconfig.ts';

test('esbuild bundling in Docker', () => {
  Bundling.bundle({
    entry,
    depsLockFilePath,
    runtime: Runtime.NODEJS_12_X,
    environment: {
      KEY: 'value',
    },
    loader: {
      '.png': 'dataurl',
    },
    forceDockerBundling: true,
  });

  // Correctly bundles with esbuild
  expect(Code.fromAsset).toHaveBeenCalledWith(path.dirname(depsLockFilePath), {
    assetHashType: AssetHashType.OUTPUT,
    bundling: expect.objectContaining({
      environment: {
        KEY: 'value',
      },
      command: [
        'bash', '-c',
        'npx esbuild --bundle "/asset-input/lib/handler.ts" --target=node12 --platform=node --outfile="/asset-output/index.js" --external:aws-sdk --loader:.png=dataurl',
      ],
      workingDirectory: '/',
    }),
  });
});

test('esbuild bundling with handler named index.ts', () => {
  Bundling.bundle({
    entry: '/project/lib/index.ts',
    depsLockFilePath,
    runtime: Runtime.NODEJS_12_X,
    forceDockerBundling: true,
  });

  // Correctly bundles with esbuild
  expect(Code.fromAsset).toHaveBeenCalledWith('/project', {
    assetHashType: AssetHashType.OUTPUT,
    bundling: expect.objectContaining({
      command: [
        'bash', '-c',
        'npx esbuild --bundle "/asset-input/lib/index.ts" --target=node12 --platform=node --outfile="/asset-output/index.js" --external:aws-sdk',
      ],
    }),
  });
});

test('esbuild bundling with tsx handler', () => {
  Bundling.bundle({
    entry: '/project/lib/handler.tsx',
    depsLockFilePath,
    runtime: Runtime.NODEJS_12_X,
    forceDockerBundling: true,
  });

  // Correctly bundles with esbuild
  expect(Code.fromAsset).toHaveBeenCalledWith('/project', {
    assetHashType: AssetHashType.OUTPUT,
    bundling: expect.objectContaining({
      command: [
        'bash', '-c',
        'npx esbuild --bundle "/asset-input/lib/handler.tsx" --target=node12 --platform=node --outfile="/asset-output/index.js" --external:aws-sdk',
      ],
    }),
  });
});

test('esbuild with Windows paths', () => {
  const osPlatformMock = jest.spyOn(os, 'platform').mockReturnValue('win32');

  Bundling.bundle({
    entry: 'C:\\my-project\\lib\\entry.ts',
    runtime: Runtime.NODEJS_12_X,
    depsLockFilePath: 'C:\\my-project\\package-lock.json',
    forceDockerBundling: true,
  });

  expect(Code.fromAsset).toHaveBeenCalledWith(expect.any(String), expect.objectContaining({
    bundling: expect.objectContaining({
      command: expect.arrayContaining([
        expect.stringContaining('/lib/entry.ts'),
      ]),
    }),
  }));

  osPlatformMock.mockRestore();
});

test('esbuild bundling with externals and dependencies', () => {
  const packageLock = path.join(__dirname, '..', 'package-lock.json');
  Bundling.bundle({
    entry: __filename,
    depsLockFilePath: packageLock,
    runtime: Runtime.NODEJS_12_X,
    externalModules: ['abc'],
    nodeModules: ['delay'],
    forceDockerBundling: true,
  });

  // Correctly bundles with esbuild
  expect(Code.fromAsset).toHaveBeenCalledWith(path.dirname(packageLock), {
    assetHashType: AssetHashType.OUTPUT,
    bundling: expect.objectContaining({
      command: [
        'bash', '-c',
        [
          'npx esbuild --bundle "/asset-input/test/bundling.test.js" --target=node12 --platform=node --outfile="/asset-output/index.js" --external:abc --external:delay',
          `echo \'{\"dependencies\":{\"delay\":\"${delayVersion}\"}}\' > /asset-output/package.json`,
          'cp /asset-input/package-lock.json /asset-output/package-lock.json',
          'cd /asset-output',
          'npm install',
        ].join(' && '),
      ],
    }),
  });
});

test('esbuild bundling with esbuild options', () => {
  Bundling.bundle({
    entry,
    depsLockFilePath,
    runtime: Runtime.NODEJS_12_X,
    minify: true,
    sourceMap: true,
    target: 'es2020',
    loader: {
      '.png': 'dataurl',
    },
    logLevel: LogLevel.SILENT,
    keepNames: true,
    tsconfig,
    metafile: true,
    banner: '/* comments */',
    footer: '/* comments */',
    forceDockerBundling: true,
    define: {
      'DEBUG': 'true',
      'process.env.KEY': JSON.stringify('VALUE'),
    },
  });

  // Correctly bundles with esbuild
  expect(Code.fromAsset).toHaveBeenCalledWith(path.dirname(depsLockFilePath), {
    assetHashType: AssetHashType.OUTPUT,
    bundling: expect.objectContaining({
      command: [
        'bash', '-c',
        [
          'npx esbuild --bundle "/asset-input/lib/handler.ts"',
          '--target=es2020 --platform=node --outfile="/asset-output/index.js"',
          '--minify --sourcemap --external:aws-sdk --loader:.png=dataurl',
          '--define:DEBUG=true --define:process.env.KEY="VALUE"',
          '--log-level=silent --keep-names --tsconfig=/asset-input/lib/custom-tsconfig.ts',
          '--metafile=/asset-output/index.meta.json --banner=\'/* comments */\' --footer=\'/* comments */\'',
        ].join(' '),
      ],
    }),
  });
});

test('Detects yarn.lock', () => {
  const yarnLock = path.join(__dirname, '..', 'yarn.lock');
  Bundling.bundle({
    entry: __filename,
    depsLockFilePath: yarnLock,
    runtime: Runtime.NODEJS_12_X,
    nodeModules: ['delay'],
    forceDockerBundling: true,
  });

  // Correctly bundles with esbuild
  expect(Code.fromAsset).toHaveBeenCalledWith(path.dirname(yarnLock), {
    assetHashType: AssetHashType.OUTPUT,
    bundling: expect.objectContaining({
      command: expect.arrayContaining([
        expect.stringMatching(/yarn\.lock.+yarn install/),
      ]),
    }),
  });
});

test('with Docker build args', () => {
  Bundling.bundle({
    entry,
    depsLockFilePath,
    runtime: Runtime.NODEJS_12_X,
    buildArgs: {
      HELLO: 'WORLD',
    },
    forceDockerBundling: true,
  });

  expect(fromAssetMock).toHaveBeenCalledWith(expect.stringMatching(/lib$/), expect.objectContaining({
    buildArgs: expect.objectContaining({
      HELLO: 'WORLD',
    }),
  }));
});

test('Local bundling', () => {
  const spawnSyncMock = jest.spyOn(child_process, 'spawnSync').mockReturnValue({
    status: 0,
    stderr: Buffer.from('stderr'),
    stdout: Buffer.from('stdout'),
    pid: 123,
    output: ['stdout', 'stderr'],
    signal: null,
  });

  const bundler = new Bundling({
    entry,
    depsLockFilePath,
    runtime: Runtime.NODEJS_12_X,
    environment: {
      KEY: 'value',
    },
  });

  expect(bundler.local).toBeDefined();

  const tryBundle = bundler.local?.tryBundle('/outdir', { image: Runtime.NODEJS_12_X.bundlingDockerImage });
  expect(tryBundle).toBe(true);

  expect(spawnSyncMock).toHaveBeenCalledWith(
    'bash',
    expect.arrayContaining(['-c', expect.stringContaining(entry)]),
    expect.objectContaining({
      env: expect.objectContaining({ KEY: 'value' }),
      cwd: '/project/lib',
    }),
  );

  // Docker image is not built
  expect(fromAssetMock).not.toHaveBeenCalled();

  spawnSyncMock.mockRestore();
});


test('Incorrect esbuild version', () => {
  getEsBuildVersionMock.mockReturnValueOnce('3.4.5');

  const bundler = new Bundling({
    entry,
    depsLockFilePath,
    runtime: Runtime.NODEJS_12_X,
  });

  const tryBundle = bundler.local?.tryBundle('/outdir', { image: Runtime.NODEJS_12_X.bundlingDockerImage });
  expect(tryBundle).toBe(false);
});

test('Custom bundling docker image', () => {
  Bundling.bundle({
    entry,
    depsLockFilePath,
    runtime: Runtime.NODEJS_12_X,
    dockerImage: DockerImage.fromRegistry('my-custom-image'),
    forceDockerBundling: true,
  });

  expect(Code.fromAsset).toHaveBeenCalledWith('/project', {
    assetHashType: AssetHashType.OUTPUT,
    bundling: expect.objectContaining({
      image: { image: 'my-custom-image' },
    }),
  });
});

test('with command hooks', () => {
  Bundling.bundle({
    entry,
    depsLockFilePath,
    runtime: Runtime.NODEJS_12_X,
    commandHooks: {
      beforeBundling(inputDir: string, outputDir: string): string[] {
        return [
          `echo hello > ${inputDir}/a.txt`,
          `cp ${inputDir}/a.txt ${outputDir}`,
        ];
      },
      afterBundling(inputDir: string, outputDir: string): string[] {
        return [`cp ${inputDir}/b.txt ${outputDir}/txt`];
      },
      beforeInstall() {
        return [];
      },
    },
    forceDockerBundling: true,
  });

  expect(Code.fromAsset).toHaveBeenCalledWith(path.dirname(depsLockFilePath), {
    assetHashType: AssetHashType.OUTPUT,
    bundling: expect.objectContaining({
      command: [
        'bash', '-c',
        expect.stringMatching(/^echo hello > \/asset-input\/a.txt && cp \/asset-input\/a.txt \/asset-output && .+ && cp \/asset-input\/b.txt \/asset-output\/txt$/),
      ],
    }),
  });
});
