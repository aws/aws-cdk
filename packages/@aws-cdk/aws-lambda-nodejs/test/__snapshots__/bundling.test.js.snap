// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`esbuild bundling with esbuild options 1`] = `
"(() => {
  // test/integ-handlers/define.ts
  function handler() {
    return [
      \\"VALUE\\",
      true,
      7777,
      'this is a \\"test\\"'
    ];
  }
})();
"
`;
